lib/elixir/lib/calendar.ex

```diff
@@ -363,4 +363,504 @@ defmodule Calendar do
+  P      | "am" or "pm" (noon is "pm", midnight as "am")                           | am, pm
+  q      | Quarter                                                                 | 1, 2, 3, 4
+  S      | Second                                                                  | 00, 59, 60
+  u      | Day of the week                                                         | 1 (Monday), 7 (Sunday)
+  x      | Preferred date (without time) representation                            | 2018-10-17
+  X      | Preferred time (without date) representation                            | 12:34:56
+  y      | Year as 2-digits                                                        | 01, 01, 86, 18
+  Y      | Year                                                                    | -0001, 0001, 1986
+  z      | +hhmm/-hhmm time zone offset from UTC (empty string if naive)           | +0300, -0530
+  Z      | Time zone abbreviation (empty string if naive)                          | CET, BRST
+  %      | Literal "%" character                                                   | %
+
+  Any other character will be interpreted as an invalid format and raise an error
+
+  ## Examples
+
+  Without options:
+
+      iex> Calendar.strftime(~U[2019-08-26 13:52:06.0Z], "%y-%m-%d %I:%M:%S %p")
+      "19-08-26 01:52:06 PM"
+
+      iex> Calendar.strftime(~U[2019-08-26 13:52:06.0Z], "%a, %B %d %Y")
+      "Mon, August 26 2019"
+
+      iex> Calendar.strftime(~U[2019-08-26 13:52:06.0Z], "%c")
+      "2019-08-26 13:52:06"
+
+  With options:
+
+      iex> Calendar.strftime(~U[2019-08-26 13:52:06.0Z], "%c", preferred_datetime: "%H:%M:%S %d-%m-%y")
+      "13:52:06 26-08-19"
+
+      iex> Calendar.strftime(
+      ...>  ~U[2019-08-26 13:52:06.0Z],
+      ...>  "%A",
+      ...>  day_of_week_names: fn day_of_week ->
+      ...>    {"segunda-feira", "terça-feira", "quarta-feira", "quinta-feira",
+      ...>    "sexta-feira", "sábado", "domingo"}
+      ...>    |> elem(day_of_week - 1)
+      ...>  end
+      ...>)
+      "segunda-feira"
+
+      iex> Calendar.strftime(
+      ...>  ~U[2019-08-26 13:52:06.0Z],
+      ...>  "%B",
+      ...>  month_names: fn month ->
+      ...>    {"январь", "февраль", "март", "апрель", "май", "июнь",
+      ...>    "июль", "август", "сентябрь", "октябрь", "ноябрь", "декабрь"}
+      ...>    |> elem(month - 1)
+      ...>  end
+      ...>)
+      "август"
+  """
+  @spec strftime(map(), String.t(), keyword()) :: String.t()
+  def strftime(date_or_time_or_datetime, string_format, user_options \\ []) do
+    parse(
+      string_format,
+      date_or_time_or_datetime,
+      options(user_options),
+      []
+    )
+    |> IO.iodata_to_binary()
+  end
+
+  defp parse("", _datetime, _format_options, acc),
+    do: Enum.reverse(acc)
+
+  defp parse("%" <> rest, datetime, format_options, acc),
+    do: parse_modifiers(rest, nil, nil, {datetime, format_options, acc})
+
+  defp parse(<<char::binary-1, rest::binary>>, datetime, format_options, acc),
```

Comment:
It should be more efficient not to create a binary here, but just an integer - it will avoid an allocation

Reply:
```suggestion
  defp parse(<<char, rest::binary>>, datetime, format_options, acc),
```